File: sdk/python/tempmail_sdk/providers/tenminute_one.py
"""
10minutemail.one：SSR __NUXT_DATA__ 中的 mailServiceToken（JWT）+ 页面 emailDomains；
收信 GET web.10minutemail.one/api/v1/mailbox/{email}
"""

import base64
import json
import re
import secrets
import time
from typing import Any, Dict, List, Optional

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import EmailInfo, Email

CHANNEL = "10minute-one"
SITE_ORIGIN = "https://10minutemail.one"
API_BASE = "https://web.10minutemail.one/api/v1"
_DEFAULT_UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
)

_NUXT_DATA_RE = re.compile(
    r'<script[^>]*\bid="__NUXT_DATA__"[^>]*>([\s\S]*?)</script>',
    re.I,
)
_JWT_RE = re.compile(r"^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$")


def _enc_mailbox_email(email: str) -> str:
    if "@" in email:
        local, _, dom = email.partition("@")
        from urllib.parse import quote

        return f"{quote(local, safe='')}%40{quote(dom, safe='')}"
    from urllib.parse import quote

    return quote(email, safe="")


def _parse_nuxt_array(html: str) -> List[Any]:
    m = _NUXT_DATA_RE.search(html)
    if not m:
        raise RuntimeError("10minute-one: __NUXT_DATA__ not found")
    return json.loads(m.group(1).strip())


def _resolve_ref(arr: List[Any], value: Any, depth: int = 0) -> Any:
    if depth > 64:
        return value
    if isinstance(value, bool):
        return value
    if isinstance(value, int) and not isinstance(value, bool) and value >= 0:
        if value < len(arr):
            return _resolve_ref(arr, arr[value], depth + 1)
    return value


def _parse_mail_service_token(arr: List[Any]) -> str:
    n = min(len(arr), 48)
    for i in range(n):
        el = arr[i]
        if isinstance(el, dict) and "mailServiceToken" in el:
            t = _resolve_ref(arr, el["mailServiceToken"])
            if isinstance(t, str) and _JWT_RE.match(t):
                return t
    for el in arr:
        if isinstance(el, dict) and "mailServiceToken" in el:
            t = _resolve_ref(arr, el["mailServiceToken"])
            if isinstance(t, str) and _JWT_RE.match(t):
                return t
    for el in arr:
        if isinstance(el, str) and _JWT_RE.match(el):
            return el
    raise RuntimeError("10minute-one: mailServiceToken not found")


def _parse_quoted_json_array(html: str, field: str) -> List[str]:
    key = f'{field}:"'
    start = html.find(key)
    if start < 0:
        return []
    slice_start = start + len(key)
    if slice_start >= len(html) or html[slice_start] != "[":
        return []
    depth = 0
    j = slice_start
    while j < len(html):
        c = html[j]
        if c == "[":
            depth += 1
        elif c == "]":
            depth -= 1
            if depth == 0:
                j += 1
                break
        j += 1
    raw = html[slice_start:j]
    unesc = raw.replace('\\"', '"')
    try:
        v = json.loads(unesc)
        return v if isinstance(v, list) else []
    except json.JSONDecodeError:
        return []


def _pick_locale(domain: Optional[str]) -> str:
    s = (domain or "").strip()
    if not s:
        return "zh"
    if "." in s and "/" not in s:
        return "zh"
    return s


def _pick_mailbox_domain(hint: Optional[str], available: List[str]) -> str:
    if not available:
        raise RuntimeError("10minute-one: no email domains")
    if hint:
        h = hint.strip().lower()
        if "." in h:
            for d in available:
                if d.lower() == h:
                    return d
    return secrets.choice(available)


def _random_local(n: int) -> str:
    alphabet = "abcdefghijklmnopqrstuvwxyz0123456789"
    return "".join(secrets.choice(alphabet) for _ in range(n))


def _jwt_exp_ms(token: str) -> Optional[int]:
    parts = token.split(".")
    if len(parts) < 2:
        return None
    b = parts[1] + "=" * (-len(parts[1]) % 4)
    try:
        pad = b.replace("-", "+").replace("_", "/")
        payload = json.loads(base64.b64decode(pad).decode("utf-8"))
        exp = payload.get("exp")
        if isinstance(exp, (int, float)):
            return int(exp * 1000)
    except (ValueError, json.JSONDecodeError, OSError):
        pass
    return None


def _api_headers(tok: str) -> Dict[str, str]:
    return {
        "Accept": "*/*",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        "Authorization": f"Bearer {tok}",
        "Cache-Control": "no-cache",
        "Content-Type": "application/json",
        "DNT": "1",
        "Origin": SITE_ORIGIN,
        "Pragma": "no-cache",
        "Referer": f"{SITE_ORIGIN}/",
        "User-Agent": _DEFAULT_UA,
        "X-Request-ID": secrets.token_hex(16),
        "X-Timestamp": str(int(time.time())),
    }


def _item_needs_detail(m: Dict[str, Any]) -> bool:
    if not str(m.get("id", "")).strip():
        return False
    subj = str(m.get("subject") or m.get("mail_title") or "").strip()
    body = (
        str(m.get("text") or m.get("body") or m.get("html") or m.get("mail_text") or "")
        .strip()
    )
    return not subj and not body


def generate_email(domain: Optional[str] = None, **kwargs) -> EmailInfo:
    loc = _pick_locale(domain)
    page_url = f"{SITE_ORIGIN}/{loc}"
    r = tm_http.get(
        page_url,
        headers={
            "User-Agent": _DEFAULT_UA,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
            "Cache-Control": "no-cache",
            "Pragma": "no-cache",
            "DNT": "1",
            "Referer": page_url,
        },
    )
    r.raise_for_status()
    html = r.text
    arr = _parse_nuxt_array(html)
    token = _parse_mail_service_token(arr)

    domains = _parse_quoted_json_array(html, "emailDomains")
    if not domains:
        domains = ["xghff.com", "oqqaj.com", "psovv.com"]

    blocked = {u.lower() for u in _parse_quoted_json_array(html, "blockedUsernames")}

    dom_hint = domain.strip() if domain and "." in domain.strip() else None
    mail_domain = _pick_mailbox_domain(dom_hint, domains)

    local = ""
    for _ in range(32):
        cand = _random_local(10)
        if cand.lower() not in blocked:
            local = cand
            break
    if not local:
        raise RuntimeError("10minute-one: could not pick username")

    address = f"{local}@{mail_domain}"
    exp_ms = _jwt_exp_ms(token)
    return EmailInfo(
        channel=CHANNEL,
        email=address,
        _token=token,
        expires_at=exp_ms,
    )


def get_emails(email: str, token: str) -> List[Email]:
    url = f"{API_BASE}/mailbox/{_enc_mailbox_email(email)}"
    r = tm_http.get(url, headers=_api_headers(token))
    r.raise_for_status()
    data = r.json()
    if not isinstance(data, list):
        raise RuntimeError("10minute-one: invalid inbox JSON")

    out: List[Email] = []
    for raw in data:
        if not isinstance(raw, dict):
            continue
        row: Dict[str, Any] = dict(raw)
        if _item_needs_detail(row):
            mid = str(row.get("id", ""))
            from urllib.parse import quote

            du = f"{API_BASE}/mailbox/{_enc_mailbox_email(email)}/{quote(mid, safe='')}"
            try:
                d = tm_http.get(du, headers=_api_headers(token))
                if d.ok:
                    detail = d.json()
                    if isinstance(detail, dict):
                        for k, v in detail.items():
                            row.setdefault(k, v)
            except OSError:
                pass
        out.append(normalize_email(row, email))
    return out

File: sdk/ruby/lib/tempmail_sdk/providers/tenminute_one.rb
# frozen_string_literal: true

require "uri"
require "json"
require "securerandom"
require "base64"
require "time"

module TempmailSdk
  module Providers
    # 10minutemail.one 渠道
    # SSR __NUXT_DATA__ 中的 mailServiceToken（JWT）+ 页面 emailDomains；
    # 收信 GET web.10minutemail.one/api/v1/mailbox/{email}
    module TenminuteOne
      CHANNEL = "10minute-one"
      SITE_ORIGIN = "https://10minutemail.one"
      API_BASE = "https://web.10minutemail.one/api/v1"
      KNOWN_EMAIL_DOMAINS = %w[xghff.com oqqaj.com psovv.com dbwot.com ygwpr.com imxwe.com].freeze
      DEFAULT_UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

      NUXT_DATA_RE = /<script[^>]*\bid="__NUXT_DATA__"[^>]*>([\s\S]*?)<\/script>/i
      JWT_RE = /\A[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\z/

      module_function

      # URL 编码邮箱（@ 编码为 %40）
      def enc_mailbox_email(email)
        if email.include?("@")
          local, _, dom = email.partition("@")
          "#{URI.encode_www_form_component(local)}%40#{URI.encode_www_form_component(dom)}"
        else
          URI.encode_www_form_component(email)
        end
      end

      # 解析 __NUXT_DATA__ JSON 数组
      def parse_nuxt_array(html)
        m = html.match(NUXT_DATA_RE)
        raise "10minute-one: __NUXT_DATA__ not found" unless m

        JSON.parse(m[1].strip)
      end

      # 解引用 NUXT 数组中的值（递归追踪索引引用）
      def resolve_ref(arr, value, depth = 0)
        return value if depth > 64
        return value if value == true || value == false
        return value unless value.is_a?(Integer) && value >= 0 && value < arr.length

        resolve_ref(arr, arr[value], depth + 1)
      end

      # 从 NUXT 数组中提取 mailServiceToken（JWT）
      def parse_mail_service_token(arr)
        # 先扫描前 48 项
        n = [arr.length, 48].min
        n.times do |i|
          el = arr[i]
          next unless el.is_a?(Hash) && el.key?("mailServiceToken")

          t = resolve_ref(arr, el["mailServiceToken"])
          return t if t.is_a?(String) && t.match?(JWT_RE)
        end
        # 全量扫描
        arr.each do |el|
          next unless el.is_a?(Hash) && el.key?("mailServiceToken")

          t = resolve_ref(arr, el["mailServiceToken"])
          return t if t.is_a?(String) && t.match?(JWT_RE)
        end
        # 最后尝试找裸 JWT 字符串
        arr.each do |el|
          return el if el.is_a?(String) && el.match?(JWT_RE)
        end
        raise "10minute-one: mailServiceToken not found"
      end

      # 从 HTML 页面中提取字段对应的 JSON 数组（如 emailDomains、blockedUsernames）
      def parse_quoted_json_array(html, field)
        key = "#{field}:\""
        start = html.index(key)
        return [] unless start

        slice_start = start + key.length
        return [] if slice_start >= html.length || html[slice_start] != "["

        depth = 0
        j = slice_start
        while j < html.length
          c = html[j]
          if c == "["
            depth += 1
          elsif c == "]"
            depth -= 1
            if depth.zero?
              j += 1
              break
            end
          end
          j += 1
        end
        raw = html[slice_start...j]
        unesc = raw.gsub('\\"', '"')
        v = JSON.parse(unesc)
        v.is_a?(Array) ? v : []
      rescue JSON::ParserError
        []
      end

      def pick_locale(domain)
        s = domain.to_s.strip
        return "zh" if s.empty?
        return "zh" if s.include?(".") && !s.include?("/")

        s
      end

      def pick_mailbox_domain(hint, available)
        raise "10minute-one: no email domains" if available.empty?

        if hint && !hint.empty?
          h = hint.strip.downcase
          if h.include?(".")
            found = available.find { |d| d.downcase == h }
            return found if found
          end
        end
        available.sample
      end

      def random_local(n)
        chars = "abcdefghijklmnopqrstuvwxyz0123456789"
        Array.new(n) { chars[rand(chars.length)] }.join
      end

      # 从 JWT 中提取过期时间戳（毫秒）
      def jwt_exp_ms(token)
        parts = token.split(".")
        return nil if parts.length < 2

        b = parts[1]
        # 补齐 Base64 填充
        b += "=" * ((-b.length) % 4)
        b = b.tr("-_", "+/")
        payload = JSON.parse(Base64.decode64(b))
        exp = payload["exp"]
        return (exp * 1000).to_i if exp.is_a?(Numeric)

        nil
      rescue StandardError
        nil
      end

      def api_headers(tok)
        {
          "Accept" => "*/*",
          "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
          "Authorization" => "Bearer #{tok}",
          "Cache-Control" => "no-cache",
          "Content-Type" => "application/json",
          "DNT" => "1",
          "Origin" => SITE_ORIGIN,
          "Pragma" => "no-cache",
          "Referer" => "#{SITE_ORIGIN}/",
          "User-Agent" => DEFAULT_UA,
          "X-Request-ID" => SecureRandom.hex(16),
          "X-Timestamp" => Time.now.to_i.to_s
        }
      end

      def item_needs_detail(m)
        return false if m["id"].to_s.strip.empty?

        body = (m["text"] || m["body"] || m["html"] || m["mail_text"] || "").to_s.strip
        body.empty?
      end

      # @param domain [String, nil] 域名或 locale 提示
      # @return [EmailInfo]
      def generate_email(domain = nil)
        loc = pick_locale(domain)
        page_url = "#{SITE_ORIGIN}/#{loc}"
        r = Http.get(page_url, headers: {
                       "User-Agent" => DEFAULT_UA,
                       "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                       "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
                       "Cache-Control" => "no-cache",
                       "Pragma" => "no-cache",
                       "DNT" => "1",
                       "Referer" => page_url
                     })
        r.raise_for_status
        html = r.body

        arr = parse_nuxt_array(html)
        token = parse_mail_service_token(arr)

        domains = (parse_quoted_json_array(html, "emailDomains") + KNOWN_EMAIL_DOMAINS).uniq
        domains = KNOWN_EMAIL_DOMAINS.dup if domains.empty?

        blocked = parse_quoted_json_array(html, "blockedUsernames").map(&:downcase).to_set

        dom_hint = domain && domain.to_s.strip.include?(".") ? domain.to_s.strip : nil
        mail_domain = pick_mailbox_domain(dom_hint, domains)

        local = nil
        32.times do
          cand = random_local(10)
          unless blocked.include?(cand.downcase)
            local = cand
            break
          end
        end
        raise "10minute-one: could not pick username" unless local

        address = "#{local}@#{mail_domain}"
        exp_ms = jwt_exp_ms(token)
        EmailInfo.new(channel: CHANNEL, email: address, token: token, expires_at: exp_ms)
      end

      # @param email [String]
      # @param token [String]
      # @return [Array<Email>]
      def get_emails(email, token)
        url = "#{API_BASE}/mailbox/#{enc_mailbox_email(email)}"
        r = Http.get(url, headers: api_headers(token))
        r.raise_for_status
        data = r.json
        raise "10minute-one: invalid inbox JSON" unless data.is_a?(Array)

        data.filter_map do |raw|
          next unless raw.is_a?(Hash)

          row = raw.dup
          if item_needs_detail(row)
            mid = row["id"].to_s
            detail_url = "#{API_BASE}/mailbox/#{enc_mailbox_email(email)}/#{URI.encode_www_form_component(mid)}"
            begin
              d = Http.get(detail_url, headers: api_headers(token))
              if d.ok?
                detail = d.json
                if detail.is_a?(Hash)
                  detail.each { |k, v| row[k] = v unless row.key?(k) }
                end
              end
            rescue StandardError
              # 忽略详情失败
            end
          end
          Normalize.normalize_email(row, email)
        end
      end
    end
  end
end
